File: python/stack.py
"""
Stack Implementation in Python
Based on Java Stack implementation
"""

from typing import TypeVar, Generic, Optional
from array import Array

T = TypeVar('T')


class Stack(Generic[T]):
    """Stack interface"""

    def size(self) -> int:
        """Get stack size"""
        raise NotImplementedError

    def is_empty(self) -> bool:
        """Check if stack is empty"""
        raise NotImplementedError

    def push(self, e: T) -> None:
        """Push element onto stack"""
        raise NotImplementedError

    def pop(self) -> T:
        """Pop element from stack"""
        raise NotImplementedError

    def peek(self) -> Optional[T]:
        """Get top element without removing"""
        raise NotImplementedError


class ArrayStack(Stack[T]):
    """Stack implementation using dynamic array"""

    def __init__(self):
        self._array: Array[T] = Array[T]()

    def size(self) -> int:
        return self._array.size()

    def is_empty(self) -> bool:
        return self._array.is_empty()

    def push(self, e: T) -> None:
        self._array.add_last(e)

    def pop(self) -> T:
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self._array.remove_last()

    def peek(self) -> Optional[T]:
        if self.is_empty():
            return None
        return self._array.get(self._array.size() - 1)

    def __str__(self) -> str:
        return f"ArrayStack: [{', '.join(str(self._array.get(i)) for i in range(self._array.size()))}] <- top"

    def __repr__(self) -> str:
        return self.__str__()


class LinkListStack(Stack[T]):
    """Stack implementation using linked list"""

    from typing import List
    from linklist import LinkList

    def __init__(self):
        self._list: LinkList[T] = LinkList[T]()

    def size(self) -> int:
        return self._list.get_size()

    def is_empty(self) -> bool:
        return self._list.is_empty()

    def push(self, e: T) -> None:
        self._list.add_first(e)

    def pop(self) -> T:
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self._list.delete_first()

    def peek(self) -> Optional[T]:
        if self.is_empty():
            return None
        return self._list.get_first()

    def __str__(self) -> str:
        return f"LinkListStack: top <- [{self._list.__str__()}]"


if __name__ == "__main__":
    # Test ArrayStack
    print("Testing ArrayStack:")
    stack = ArrayStack[int]()
    for i in range(5):
        stack.push(i)
        print(f"Push {i}: {stack}")

    print(f"Peek: {stack.peek()}")
    print(f"Pop: {stack.pop()}")
    print(f"Pop: {stack.pop()}")
    print(f"After pops: {stack}")

    # Test LinkListStack
    print("\nTesting LinkListStack:")
    ll_stack = LinkListStack[int]()
    for i in range(5):
        ll_stack.push(i)
        print(f"Push {i}: {ll_stack}")

    print(f"Peek: {ll_stack.peek()}")
    print(f"Pop: {ll_stack.pop()}")
    print(f"After pop: {ll_stack}")

File: golang/Stack/stack.go
package main

import (
	"fmt"
	"strings"
)

// Stack 栈接口
type Stack[E any] interface {
	Size() int
	IsEmpty() bool
	Push(e E)
	Pop() E
	Peek() E
}

// ArrayStack 基于动态数组的栈实现
type ArrayStack[E any] struct {
	array *Array[E]
}

// NewArrayStack 创建一个新的数组栈
func NewArrayStack[E any](capacity int) *ArrayStack[E] {
	return &ArrayStack[E]{
		array: NewArray[E](capacity),
	}
}

// NewArrayStackDefault 创建默认容量的数组栈
func NewArrayStackDefault[E any]() *ArrayStack[E] {
	return &ArrayStack[E]{
		array: NewArrayDefault[E](),
	}
}

func (s *ArrayStack[E]) Size() int {
	return s.array.Size()
}

func (s *ArrayStack[E]) IsEmpty() bool {
	return s.array.IsEmpty()
}

func (s *ArrayStack[E]) Push(e E) {
	s.array.AddLast(e)
}

func (s *ArrayStack[E]) Pop() E {
	if s.IsEmpty() {
		panic("Stack is empty")
	}
	return s.array.RemoveLast()
}

func (s *ArrayStack[E]) Peek() E {
	if s.IsEmpty() {
		panic("Stack is empty")
	}
	return s.array.Get(s.array.Size() - 1)
}

func (s *ArrayStack[E]) String() string {
	var builder strings.Builder
	builder.WriteString("Stack: [")
	for i := 0; i < s.array.Size(); i++ {
		builder.WriteString(fmt.Sprintf("%v", s.array.Get(i)))
		if i < s.array.Size()-1 {
			builder.WriteString(", ")
		}
	}
	builder.WriteString("] <- top")
	return builder.String()
}

func main() {
	fmt.Println("=== Go语言栈示例 ===")

	stack := NewArrayStackDefault[int]()
	for i := 0; i < 5; i++ {
		stack.Push(i)
		fmt.Printf("入栈 %d: %s\n", i, stack.String())
	}

	fmt.Printf("栈顶元素: %d\n", stack.Peek())
	fmt.Printf("出栈: %d\n", stack.Pop())
	fmt.Printf("出栈后: %s\n", stack.String())

	fmt.Printf("栈大小: %d\n", stack.Size())
	fmt.Printf("是否为空: %t\n", stack.IsEmpty())

	// 清空栈
	for !stack.IsEmpty() {
		stack.Pop()
	}
	fmt.Printf("清空后是否为空: %t\n", stack.IsEmpty())

	fmt.Println("\n=== 示例程序结束 ===")
}
